content/Refactoring - Sebastian Herrera.md

```diff
@@ -0,0 +1,146 @@
+- If you struggle to understand the code, refactor it so you won't have to struggle again next time you look at it.
+
+**Conclusion: Refactoring is a part of day-to-day programming :calendar:**
+
+## 3) CODE SMELLS
+Probably you’ve heard about code smells, and for sure you have already seen tons of them. Let’s see their characteristics.
+- A code smell is a surface indication that usually corresponds to a deeper problem in the system.
+  - Firstly a smell is by definition something quick to spot - or sniffable.
+  - A long method is a good example of this - just looking at the code and my nose twitches if I see more than a dozen lines.
+- The second is that smells don't always indicate a problem.
+  - Some long methods are just fine. You have to look deeper to see if there is an underlying problem - smells aren't inherently bad on their own - they are often an indicator of a problem rather than the problem themselves.
+
+Below is an example written in Ruby (it should be simple enough for anyone to follow). Don't look at the image yet! I know the gray boxes already grab much attention, just cover them with your hand and look at the code for no more than 10 seconds, thinking of the code smells that you can find. Then uncover the boxes. I’m quite sure that in that amount of time you could find the same code smells (or even more). What I’m trying to show, is exactly what code smells are, they are something very quick to spot, and they are not a problem by themselves (think that this code works perfectly fine).
+
+![Code Smells Example](images/code_smells_example.png)
+
+Here you can find a [catalog of code smells](https://github.com/lee-dohm/code-smells/blob/master/README.md), defined mainly by Martin Fowler and Kent Beck.
+
+I highly recommend reading, understanding and memorizing them. If you do, whenever you are doing code review to another person, or just reviewing your code, you will see much more easily the possible improvements.
+
+**Conclusion: Pay attention to code smells. They might be whispering (or even shouting!) asking for a better solution :warning: :eyes:**
+
+## 4) ROLE OF TESTING
+By definition, refactoring is the process of changing the code structure, without altering its behavior.
+- How do we know that we have not altered its behavior then?
+  - Yes, you’ve guessed. We need tests.
+- And what if the code we want to refactor doesn’t have enough tests?
+  - Then you should first write the tests, and then refactor.
+
+You may be asking other things like whether it's better to have more unit tests or integration tests, or if it's necessary to have near 100% code coverage on tests.
+
+The short answer is, there is no specific technique or condition to be met in order to be sure that your code works exactly as expected. There is a quote from Dijkstra that says *"Program testing can be used to show the presence of bugs, but never to show their absence."*
```

Comment:
Maybe can rewrite this. And not sure about "in order to be sure that your code works exactly as expected"